### PR TITLE
feat: Update registry fallbacks

### PR DESCRIPTION
**Pull request by simar7**
## Description

Adds public ECR registry as a fallback for both vuln-db and java-db. 

## Related PRs
- [x] https://github.com/aquasecurity/trivy/pull/7605/

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by knqyf263** [comment]
We should document how to authenticate with ECR. Otherwise, ECR Public has only 500 GB free tier for anonymous users. Also, it's calculated by IP address. I suppose we'll run out of the free tier on CI service as it shares IP addresses.

**Comment by DmitriyLewen** [comment]
I'm worried about adding `ecr` as the default registry.
I think most users will ignore the authentication recommendation, and some users just don't read the docs.

Instead of adding `ecr` - I suggest adding warning if downloading db from `ghcr` returns error.
We will show information about using `ecr` (with authentication information) + link to the docs.

**Comment by nvuillam** [comment]
We read the docs but... having to authenticate to aws would be a breaking change 🥹

**Comment by DmitriyLewen** [comment]
hello @nvuillam 
Authentication is necessary to preserve aws limits for unauthenticated users as much as possible.
But we don't require it. We just ask users to do it if possible - it will help avoid rate limiting errors like ghcr.io.

File: docs/docs/configuration/db.md [hunk 1]
```diff
@@ -92,6 +92,21 @@ You can reference the OCI manifest of [trivy-db].
...
+Trivy user, there are certain recommendations that one can make in order to ensure rate limits are not hit.
+
+#### Authenticated use of Registries
+By authenticating with the registries that Trivy hosts its DBs on can significantly increase the limit for users. For Amazon ECR, the details for rate limits can be found [ecr-limits].
+For GitHub GHCR, the rate limits can be found [ghcr-limits]
+
+Please see more info on how to authenticate with ECR [auth-ecr] and GHCR [auth-ghcr].
+
+#### Caching DBs
+Trivy DB and Trivy Java DB are published every 24 hours. If you are running Trivy scans more often than this, you can significantly benefit from caching the DBs on each run and updating them as needed.
```
Comment (DmitriyLewen) [inline docs/docs/configuration/db.md:107]: 24 hours is incorrect.
We publish trivy-db every 6 hours, but update interval is 24 hours.
for trivy-java-db: publish every 24 hours, update interval is 3 days.

**Comment by knqyf263** [comment]
>We will show information about using ecr (with authentication information) + link to the docs.

I'm also leaning toward this approach. I'm afraid that users downloading container images from ECR Public (e.g., Kubernetes cluster) may suddenly reach rate limits due to Trivy DB and be unable to deploy, etc.

Of course, it is best if users can continue to use DBs as before without being aware of it, but the implicit use of ECR Public may have other negative effects. We should be careful about the use of ECR Public. If users explicitly set this up, they will understand these trade-offs.

**Comment by simar7** [comment]
> > We will show information about using ecr (with authentication information) + link to the docs.
> 
> I'm also leaning toward this approach. I'm afraid that users downloading container images from ECR Public (e.g., Kubernetes cluster) may suddenly reach rate limits due to Trivy DB and be unable to deploy, etc.
> 

Yes but the same can be said for GHCR today. 

> Of course, it is best if users can continue to use DBs as before without being aware of it, but the implicit use of ECR Public may have other negative effects. We should be careful about the use of ECR Public. If users explicitly set this up, they will understand these trade-offs.

I added some docs [here](https://github.com/aquasecurity/trivy/blob/fa031898ebd2d39009e6bace47b288595793a6d7/docs/docs/configuration/db.md#rate-limits).

Ultimately we can reach rate limits even with authenticated use of container registries. The fallback to using a secondary registry only mitigates the risk up to an extent.

**Comment by knqyf263** [comment]
>Yes but the same can be said for GHCR today.

The rate limits in GHCR are on the organization, not the user, unlike ECR Public. In other words, heavy use of GHCR may make aquasecurity images unavailable, but not other images. Users may not be able to use Trivy, but they will be able to pull any other images except aquasecurity.

**Comment by goatwu1993** [comment]
checks bundle also needs ecr. 

```yaml
        env:
          TRIVY_DB_REPOSITORY: public.ecr.aws/aquasecurity/trivy-db
          TRIVY_JAVA_DB_REPOSITORY: public.ecr.aws/aquasecurity/trivy-java-db
          TRIVY_CHECKS_BUNDLE_REPOSITORY: public.ecr.aws/aquasecurity/trivy-checks
```

**Comment by bgoareguer** [comment]
The `trivy.dbRepository` parameter in the `helm/trivy/values.yaml` file also needs to be updated.

**Comment by afdesk** [comment]
> The `trivy.dbRepository` parameter in the `helm/trivy/values.yaml` file also needs to be updated.

thanks for the good idea.
it should be add after the PR is merged.

**Comment by afdesk** [comment]
@simar7 
Now trivy-db pushes the database to Docker Hub: https://github.com/aquasecurity/trivy-db/pull/448
maybe it makes sense to add Docker Hub as a default repository too?
just idea. thanks!

**Comment by jeanmarcmoeckel** [comment]
Any updates here?

**Comment by knqyf263** [comment]
`mirror.gcr.io` seems the best for our use case.
https://github.com/aquasecurity/trivy/discussions/7538#discussioncomment-11258881

**Comment by benji78** [comment]
I see 8 incorrect mentions of `public.ecr.aws/aquasecurity/trivy-db:1` which should be `trivy-java-db` !

File: pkg/javadb/client.go [hunk 2]
```diff
@@ -29,6 +29,9 @@ const (
 var (
 	// GitHub Container Registry
 	DefaultGHCRRepository = fmt.Sprintf("%s:%d", "ghcr.io/aquasecurity/trivy-java-db", SchemaVersion)
+
+	// AWS ECR
+	DefaultECRRepository = fmt.Sprintf("%s:%d", "public.ecr.aws/aquasecurity/trivy-db", SchemaVersion)
```
Comment (DmitriyLewen) [inline pkg/javadb/client.go:34]:
```suggestion
	DefaultECRRepository = fmt.Sprintf("%s:%d", "public.ecr.aws/aquasecurity/trivy-java-db", SchemaVersion)
```

Comment (DmitriyLewen) [reply pkg/javadb/client.go:34]: Thanks @benji78 !

File: pkg/remote/remote.go [hunk 3]
```diff
@@ -154,10 +154,14 @@ func authOptions(ctx context.Context, ref name.Reference, option types.RegistryO
 		}))
 	}
 
-	domain := ref.Context().RegistryStr()
-	token := registry.GetToken(ctx, domain, option)
-	if !lo.IsEmpty(token) {
-		opts = append(opts, remote.WithAuth(&token))
+	// The default trivy-db and trivy-java-db repositories are public.
```
Comment (knqyf263) [inline pkg/remote/remote.go:158]: If I understand correctly, there is no chance that `mirror.gcr.io` requires authentication. We should skip `mirror.gcr.io` here.
https://github.com/aquasecurity/trivy/blob/94791f83431c02a566f73e4e3f00557d3e90aef5/pkg/fanal/image/registry/google/google.go#L32

Comment (DmitriyLewen) [reply pkg/remote/remote.go]: yeah, you're probably right, I didn't think about it
Changed in https://github.com/aquasecurity/trivy/pull/7679/commits/99e96bbd53777740fc0957211438ae336fff72d3

File: docs/docs/configuration/db.md [hunk 4]
```diff
@@ -92,6 +92,20 @@ You can reference the OCI manifest of [trivy-db].
     `trivy-db-registry:latest` => `trivy-db-registry:latest`, but `trivy-db-registry` => `trivy-db-registry:2`.
 
 
+### Rate limits
+Trivy hosts its databases on public OCI registries that are subject to their respective rate limits. While we strive to make the databases available to every
+Trivy user, there are certain recommendations that one can make in order to ensure rate limits are not hit.
+
+#### Authenticated use of Registries
+By authenticating with the registries that Trivy hosts its DBs on can significantly increase the limit for users. For Amazon ECR, the details for rate limits can be found [ecr-limits].
+For GitHub GHCR, the rate limits can be found [ghcr-limits]
```
Comment (knqyf263) [inline docs/docs/configuration/db.md:101]: It's not correct. GHCR doesn't respect the GitHub token now.

File: pkg/javadb/client.go [hunk 5]
```diff
@@ -29,6 +29,9 @@ const (
 var (
 	// GitHub Container Registry
 	DefaultGHCRRepository = fmt.Sprintf("%s:%d", "ghcr.io/aquasecurity/trivy-java-db", SchemaVersion)
+
+	// GHC mirrors
```
Comment (knqyf263) [inline pkg/javadb/client.go:33]: nit

```suggestion
	// GCR mirrors
```

File: docs/docs/configuration/db.md (hunk 4 above)
Comment (DmitriyLewen) [reply docs/docs/configuration/db.md:101]: But some users wrote that GH token helps them.
Do you think that for GHCR it makes no difference at all whether the user is authorized or not when downloading the image?

Comment (knqyf263) [reply docs/docs/configuration/db.md:101]: >But some users wrote that GH token helps them.

Trivy has [a bug](https://github.com/aquasecurity/trivy/pull/7580) that doesn't reflect `GITHUB_TOKEN` now, so I'm sure those users luckily didn't see rate limit errors.

I saw [the same rate limit error](https://github.com/aquasecurity/trivy/pull/7580#issuecomment-2370624673) when I set the token, so I believe it has no effect on GHCR. If we confirm it helps, we can add it back.

Comment (DmitriyLewen) [reply docs/docs/configuration/db.md]: Got it!
Removed info about GHCR in https://github.com/aquasecurity/trivy/pull/7679/commits/c18a51169c103016f73c8ce61a86a9c191109fb8

File: pkg/javadb/client.go (hunk 5 above)
Comment (DmitriyLewen) [reply pkg/javadb/client.go]: thanks! fixed in https://github.com/aquasecurity/trivy/pull/7679/commits/272b1a04ff23dcba2874d2835a4c2cc8b1ab3e15

**Review by knqyf263 — APPROVED** [review]

File: docs/docs/configuration/db.md [hunk 6]
```diff
@@ -92,6 +92,19 @@ You can reference the OCI manifest of [trivy-db].
...
 
 
+### Rate limits
+Trivy hosts its databases on public OCI registries that are subject to their respective rate limits. While we strive to make the databases available to every
+Trivy user, there are certain recommendations that one can make in order to ensure rate limits are not hit.
+
+#### Authenticated use of Registries
+By authenticating with the registries that Trivy hosts its DBs on can significantly increase the limit for users. For Amazon ECR, the details for rate limits can be found [ecr-limits].
+
+Please see more info on how to authenticate with ECR [auth-ecr].
```
Comment (benji78) [inline docs/docs/configuration/db.md:102]: Is Amazon ECR still relevant here if it has been replaced by GCR?

File: docs/docs/configuration/db.md [hunk 7]
```diff
@@ -92,6 +92,19 @@ You can reference the OCI manifest of [trivy-db].
...
+
+#### Caching DBs
+Trivy DB and Trivy Java DB are published every 6 hours and 24 hours, respectively. If you are running Trivy scans more often than this, you can significantly benefit from caching the DBs on each run and updating them as needed.
+Once example of this can be seen in Trivy Action, where with caching multiple CI invocations can be performed with a single download of the DBs. More on info Trivy Action caching can be found [trivy-action-cache].
```
Comment (benji78) [inline docs/docs/configuration/db.md:106]: There is an extra `c` in `Once` which should be `One example of this …`

**Review by benji78 — CHANGES_REQUESTED** [review]

File: docs/docs/configuration/db.md (hunk 6 above)
Comment (DmitriyLewen) [reply docs/docs/configuration/db.md:102]: We still push trivy-db and trivy-java-db into ECR.
So this is still relevant for users using ECR.

File: docs/docs/configuration/db.md (hunk 7 above)
Comment (DmitriyLewen) [reply docs/docs/configuration/db.md:106]: Thanks! We will fix that in another PR.

**Comment by DmitriyLewen** [comment]
@aqua-bot backport release/v0.57

**Pull request merged**